core/domain/user_jobs_one_off_test.py

```diff
@@ -377,3 +380,79 @@ def test_adding_exploration_to_collection(self):
+
+
+class UserFirstContributionMsecOneOffJobTests(test_utils.GenericTestBase):
+
+    EXP_ID = 'test_exp'
+
+    def setUp(self):
+        super(UserFirstContributionMsecOneOffJobTests, self).setUp()
+
+    def test_contribution_msec_updates_on_published_explorations(self):
+        self.signup(self.ADMIN_EMAIL, self.ADMIN_USERNAME)
+        self.admin_id = self.get_user_id_from_email(self.ADMIN_EMAIL)
+
+        exploration = self.save_new_valid_exploration(
+            self.EXP_ID, self.admin_id, end_state_name='End')
+        self.init_state_name = exploration.init_state_name
+
+        # Test that no contribution time is set.
+        job_id = (
+            user_jobs_one_off.UserFirstContributionMsecOneOffJob.create_new())
+        user_jobs_one_off.UserFirstContributionMsecOneOffJob.enqueue(job_id)
+        self.process_and_flush_pending_tasks()
+        self.assertIsNone(
+            user_services.get_user_settings(self.admin_id).first_contribution_msec)
+
+        # Test all owners and editors of exploration after publication have
+        # updated times.
+        exp_services.publish_exploration_and_update_user_profiles(
+            self.admin_id, self.EXP_ID)
+        rights_manager.release_ownership_of_exploration(
+            self.admin_id, self.EXP_ID)
+        self.signup(self.EDITOR_EMAIL, self.EDITOR_USERNAME)
+        self.editor_id = self.get_user_id_from_email(self.EDITOR_EMAIL)
+        exp_services.update_exploration(
+            self.editor_id, self.EXP_ID, [{
+                'cmd': 'edit_state_property',
+                'state_name': self.init_state_name,
+                'property_name': 'widget_id',
+                'new_value': 'MultipleChoiceInput'
+            }], 'commit')
+        job_id = (
+            user_jobs_one_off.UserFirstContributionMsecOneOffJob.create_new())
+        user_jobs_one_off.UserFirstContributionMsecOneOffJob.enqueue(job_id)
+        self.process_and_flush_pending_tasks()
+        self.assertIsNotNone(user_services.get_user_settings(self.admin_id)
+            .first_contribution_msec)
+        self.assertIsNotNone(user_services.get_user_settings(self.editor_id)
+            .first_contribution_msec)
+
+    def test_contribution_msec_does_not_update_on_unpublished_explorations(self):
+        self.signup(self.ADMIN_EMAIL, self.ADMIN_USERNAME)
+        self.admin_id = self.get_user_id_from_email(self.ADMIN_EMAIL)
+        self.set_admins([self.ADMIN_EMAIL])
+
+        self.signup(self.OWNER_EMAIL, self.OWNER_USERNAME)
+        self.owner_id = self.get_user_id_from_email(self.OWNER_EMAIL)
+
+        exploration = self.save_new_valid_exploration(
+            self.EXP_ID, self.owner_id, end_state_name='End')
+        self.init_state_name = exploration.init_state_name
+        exp_services.publish_exploration_and_update_user_profiles(
+            self.owner_id, self.EXP_ID)
+        # User's first_contribution_msec is reset to None after publish_exploration
```

Comment:
Reword to: "We now manually reset the user's first_contribution_msec to None", in order to make clear that this is an action on our part (rather than one that the system somehow does).